python/tk_multi_launchapp/software_entity_launcher.py

```diff
@@ -103,102 +198,81 @@ def register_launch_commands(self):
 
-        # Record how many Software entities were found and what the resolved
-        # filter looks like.
-        self._tk_app.log_debug("Found (%d) %s entities matching filters:\n%s " %
-            (len(sw_entities), sw_entity, pprint.pformat(sw_filters, indent=4))
-        )
-        for sw_entity in sw_entities:
-            self._tk_app.log_debug("Software Entity:\n%s" % pprint.pformat(sw_entity, indent=4))
-
-            app_menu_name = sw_entity["code"]
-            app_icon = sw_entity["image"]
-            app_engine = sw_entity["sg_engine"]
-            app_path = sw_entity[app_path_field]
-            app_args = sw_entity[app_args_field] or ""
+    def _build_register_commands(self, display_name, icon, engine, path, args, versions=None):
+        """
+        Determine the list of command data to register basded on the input
+        path and versions information.
 
-            if not app_path:
-                # If no path has been set for the app, we will eventually go look for one,
-                # but for now, don't load the app.
-                self._tk_app.log_warning("No path specified for app [%s]." % app_menu_name)
-                continue
+        :param display_name: Label string for the registered command.
+        :param icon: Path to icon to load for the registered command.
+        :param engine: Name of the Toolkit engine this command will run
+        :param path: Path to the DCC execuable to register a launch command for.
+        :param args: Args to pass to the DCC executable when launched.
+        :param versions: (optional) List of specific versions (as strings) to
+                         register launch commands for.
 
-            if app_engine:
-                # Try to retrieve the path to the specified engine. If nothing is
-                # returned, then this engine hasn't been loaded in the current
-                # environment, and there's not much more we can do.
-                app_engine_path = sgtk.platform.get_engine_path(
-                    app_engine, self._tk_app.sgtk, self._tk_app.context
-                )
-                if not app_engine_path:
-                    self._tk_app.log_warning(
-                        "Software engine %s is not loaded in the current environment. "
-                        "Cannot launch %s" % (app_engine, app_path)
-                    )
-                    continue
-
-            # Download the thumbnail to use as the app's icon.
-            if app_icon:
-                sg_icon = shotgun_data.ShotgunDataRetriever.download_thumbnail(
-                    app_icon, self._tk_app
+        :returns: List of dictionaries containing required information to register
+                  a command with the current engine.
+        """
+        # List of command data to return
+        commands = []
+        if path:
+            # Register a command for each version for the path specified.
+            for version in (versions or [None]):
```

Comment:
this seems complicated. Wonder if its easier to read if we break it into an if clause?